Added proposal to improve support for packages signed with a Trusted Signing certificate. (#13791)
NuGet/Home: accepted/2024/improve-support-for-packages-signed-with-trusted-signing.md

# Improve support for NuGet packages that are signed with a Trusted Signing certificate

- Author Name: Dirk Lemstra [dlemstra](https://github.com/dlemstra)
- GitHub Issue: https://github.com/NuGet/NuGetGallery/issues/10027

## Summary

This proposal is about improving the support for NuGet packages that are signed with a Trusted Signing certificate. This proposal
consists of three steps to improve the experience for users that sign their packages with Trusted Signing. These three steps can
be released separately but I can also understand that the NuGet team would want to launch this as a complete feature.

## Motivation

Earlier this year [Trusted Signing](https://learn.microsoft.com/en-us/azure/trusted-signing/) was launched by Microsoft and 

> recently support for this was added to `dotnet sign`. Packages signed with this can be uploaded to the NuGet Gallery. But

Reviewer: The signing tool name isn't `dotnet sign`, it's `Sign CLI`. Would probably be useful to have it as a link to the repo.

Author: I have corrected the name and included a link.

> because the certificate is only valid for three days a user will probably always need to update their certificate on their 

Reviewer: It's three days now but it could be shorter--do we want to say a specific number or say "short".

Author: I have updated it to "a short time" now.

account page. I would like to propose a set of changes to improve the experience for users that sign their NuGet package
with Trusted Signing.

## Explanation

### Functional explanation

> ### Step 1: Register EKU of certificate that was added by Trusted Signing

Reviewer 2: I think we can implement and deploy each step individually. I am concerned that Step 1 has a part this is not self-service (the deletion, potentially increasing on-call/support load) and Step 2 requires the user to extract the .cer themselves somehow. This is fine, but we can limit exposure but putting the new behavior behind a flight -- this is a feature flag but has a username filter. This means we can add specific people (perhaps MVPs who ask) to the feature until we get to step 3 and then we can open it up to everyone.

What do you think?

Author: Sounds like a good plan. Not sure if we should limit this to just MVP's though. There are probably also other people that would want to try out this feature.

Reviewer 2: Yes, I take MVPs as an example because it's easy to have tight, proactive communication with them. But certainly, we can definitely open it up to anyone that asks. Maybe a small announcement with "contact us if you want to try it", or perhaps a set of people Ian knows. Initial adoption efforts can be TBD. We can add any username to the flight system with ease.

#### How would this be implemented?

The first step would be to store the EKU that is inside the certificate that was used to sign the NuGet package with Trusted

> Signing. The certificate of Tursted Signing is only valid for thee days but it contains an EKU to that can be used to get the

Reviewer 2: ```suggestion
Signing. The certificate of Trusted Signing is only valid for thee days but it contains an EKU to that can be used to get the
```

> Signing. The certificate of Trusted Signing is only valid for thee days but it contains an EKU to that can be used to get the

Reviewer 3: ```suggestion
Signing. The certificate of Trusted Signing is currently only valid for three days, but it contains an EKU to that can be used to get the
```

identity of the owner of the certificate. There are two EKU's in the certificate. One that tells us that it was signed with
Trusted Signing (`1.3.6.1.4.1.311.97.1.0`) and another one that contains the Public Trust identity (e.g. 
`1.3.6.1.4.1.311.97.990309390.766961637.194916062.941502583`). More details about this can be found here:
https://learn.microsoft.com/en-us/azure/trusted-signing/concept-trusted-signing-cert-management. 

I would like to propose to link EKU's to the user in a similar way as that is now done with certificates:

> ![EKU](images/trusted-signing-eku.png)

Reviewer 3: I assume that this screenshot is from Trusted Signing website, but I don't know for sure.  It does not look like a mockup for NuGet.org.  I think it's important to call out that this is a screenshot from the Trusted Signing website not exactly what you're proposing for NuGet.org.

Author: This is a mermaid diagram? Are you referring to another image?

Author: I have update the text around the Azure Portal screenshot to make this more clear.

> But his would require a user to enter that EKU in the interface to link it with their account. We could also start with

Reviewer 2: ```suggestion
But this would require a user to enter that EKU in the interface to link it with their account. We could also start with
```

automatically linking the EKU to the user by using the current "certificate upload" interface. The user would need to do the
following:
- Extract the `.cer` file of the signing certificate from the NuGet package.
- Upload the `.cer` file and link it to their account.
- Upload the NuGet package that was signed with Trusted Signing.

Below is a picture that show how the certificate and EKU would be linked to user before and after the NuGet package is uploaded.

![EKU link](images/trusted-signing-eku-link.png)

This would mean that the certificate is no longer linked to the user and will no longer be shown in the interface because the EKU
is now linked to the user. That will mean that the user would need to contact the NuGet team to remove the EKU link until updates
to the interface have been made. But I don't think a lot of people will need this functionality right away.

The EKU will only be linked when the following conditions are true:

> - The package was signed within the last X days where X should be decided by the NuGet team (not sure if this is really necessary).

Reviewer 2: I spoke to the team internally. Let's make this value configurable (an appsetting.json value for the related validation job) and set the default value to 30 days. We can change this to a larger number if needed in the future, based on customer feedback.

We selected 30 days based on this data:

99.4% of packages on NuGet.org today have an author sign timestamp within 30 days of the upload time. 
@JonDouglas [ran a poll on Twitter](https://x.com/nuget/status/1836452574452498890) for this question and found this value will work for 93.5% of respondents.

Author: I have updated the window to 30 days.

> - The signing certificate must be issued by the CA certificate `Microsoft Identity Verification Root Certificate Authority 2020`.

Reviewer 3: @dlemstra, @ianjmcm's feedback [here](https://github.com/NuGet/Home/pull/13791/files#r1763442640) was that the signing certificate's issuer should _not_ be the root CA.  Certainly, the signing certificate should chain to this root, but that's not the same thing.  The signing certificate's issuer should chain to the "Microsoft ID Verified Code Signing PCA 2021" intermediate CA (or policy CA), like I've illustrated below.

I don't know if there is a hard restriction on chain length.  In the abstract, there could be not just 1 but more than one intermediate CA's between the PCA and the subscriber certificate.  @ianjmcm's point was that we can expect that the subscriber certificate will _eventually_ chain up to "Microsoft ID Verified Code Signing PCA 2021".  I think it's safest to not hard-code expectations in our code as to the chain length (i.e.:  whether there is 1 or more CA's between the PCA and the end certificate).

```mermaid
graph TD
    A["Microsoft Identity Verification Root Certificate Authority 2020"]
    click A "https://crt.sh/?q=5367f20c7ade0e2bca790915056d086b720c33c1fa2a2661acf787e3292e1270" "Root CA"
    A --> B["Microsoft ID Verified Code Signing PCA 2021"]
    click B "https://crt.sh/?q=3d29798cc5d3f0644a7e0dc9cb1cade523ea5ec83b335109b605bfeaa7d5f5c1" "Intermediate CA"
    B --> C["issuing CA (e.g.:  Microsoft ID Verified CS AOC CA 02"]
    C --> D["subscriber certificate"]
```

Author: I think we can only check if the root certificate is the `Microsoft Identity Verification Root Certificate Authority 2020` certificate and not "care about" the chain? Or am I misunderstanding you @dtivel?

Reviewer 3: @dlemstra, ianmcm said:

> Please consider that the signing certs will not be signed by the root directly, Microsoft Identity Verification Root Certificate Authority 2020, but its issuer will be chained to the Microsoft ID Verified Code Signing PCA 2021 that is signed by the root. Only code signing certs under this CA in the hierarchy should be accepted for NuGet author signing from Trusted Signing.

So,

1.  The root CA won't issue the signing certificate.
2.  We must verify that the author-signing certificate chains up to the `Microsoft ID Verified Code Signing PCA 2021` certifcate.

  (https://learn.microsoft.com/en-us/azure/trusted-signing/concept-trusted-signing-trust-models)

> - The signing certificate has a valid counter-signature (timestamp).

Reviewer 3: This is nonsensical.  There isn't standard way to countersign (or timestamp) a certificate.

I think what you meant is that the package signature has a valid timestamp.

Reviewer 3: Actually, I think you should avoid restating unchanged requirements like this.  It creates confusion around what the authoritative, complete requirements are.

[This public document](https://learn.microsoft.com/nuget/reference/signed-packages-reference#signature-requirements-on-nugetorg) already states this requirement.

You can remove any requirement here that is already in that list.

Author: The reason this requirement was included is to make sure that the EKU is one that was added by the Trusted Signing service in Azure.

Reviewer 3: > - The signing certificate has a valid counter-signature (timestamp).

This requirement no relevance to ensuring "that the EKU is one that was added by the Trusted Signing service in Azure."

The [next line](https://github.com/NuGet/Home/pull/13791/files#diff-0d12e468a60ee40833fe78c37e044eb123d6abd72a8577e6a0b3fae81647453cR59) expresses the requirement you describe.

You should remove this requirement.

Author: I have updated the proposal to remove this requirement

- The signing certificate contains the Trusted Signing EKU.
- The signing certificate contains a Public Trust Identity EKU.

> - The Public Trust Identity EKU is not already linked to the user.

Reviewer 2: To be clear, we will allow an EKU to be linked to multiple users/orgs right? This is possible today since the same .cer file can be uploaded by multiple users and this means the same fingerprint is possible. 

I think it is okay to allow this because a) both users/orgs would need access to signing permission for this to matter and b) there may be legitimate scenarios for this, such as one signing identity (TS account) but multiple NuGet.org profiles (one user with multiple orgs).

Reviewer 4: The unique EKU for the validated identity in Trusted Signing is tied to the Subscription, so other accounts and certificate profiles will be able to use that unique EKU. This allows orgs to organize their account resources with their business/development units as they see fit. 

Please consider that the signing certs will not be signed by the root directly, Microsoft Identity Verification Root Certificate Authority 2020, but its issuer will be chained to the Microsoft ID Verified Code Signing PCA 2021 that is signed by the root. Only code signing certs under this CA in the hierarchy should be accepted for NuGet author signing from Trusted Signing.

Author: I have updated the PR to make it more clear that it should be the root CA.

they uploaded earlier. This could confuse the user so we could keep the link and let the user remove it manually. But doing this
would also not make it clear that an EKU was linked to their account. Making this visible in the interface will be the next step
of these changes.

### Step 2: Change how Trusted Signing EKU's are shown in the NuGet Gallery.

#### How would this be implemented?

The next steps would be to change how this will be displayed inside the NuGet Gallery. Below is an image of how this could be
added to the account page of a user (please forgive me for my css skills). The naming is a proposal and I am open to suggestions
but I think it would be best to show the EKU's in a separate table.

![User profile](images/trusted-signing-profile.png)

This image demonstrates how this would look when a user has both an EKU and certificate linked to their account. As mentioned in
step 1 the certificate would no longer be shown and the user would only see their EKU's. The Subject and Issuer will be extracted
from the certificate the first time and added to the EKU data. Uploads in the future could result in a new Issuer so this should

> be updated but a change of Subject would result in a different EKU so there is no need to update this.

Reviewer 4: With Trusted Signing, the change in subjectDN will not result in a different EKU if the organization/individual legally changes their name and validates that new name with their existing Trusted Signing resources. If they choose to use a new subscription and identity validation for the new legal name, a new unique EKU would result.

Author: Thanks for mentioning this. Seems that I misunderstood you earlier. I have updated the PR.

#### How does this help the user?

When only step 1 is implemented it will already be a huge help for users that use Trusted Signing but it will look strange inside
the NuGet Gallery because no certificates are shown anymore and they cannot unlink their EKU. Adding this separate table with EKU's
linked to the user their account would allow them to remove this EKU. I personally don't think users would really need this
functionality right away so that is why I think this could be added as a second step.

### Step 3: Allow a user to submit a Public Trust identity EKU in the NuGet Gallery

#### How would this be implemented?

> When a `.cer` file is uploaded only the tumbprint is stored because the certificate is not parsed by the NuGet Gallery due to

Reviewer 2: ```suggestion
When a `.cer` file is uploaded only the thumbprint is stored because the certificate is not parsed by the NuGet Gallery due to
```

security reasons. This means we cannot add the Public trust identity EKU record when a certificate is uploaded. For a user that
has access to the Azure Portal it would be easier to copy the EKU from the interface:

> ![Azure Porta;](images/trusted-signing-azure.png)

Reviewer 2: ```suggestion
![Azure Portal](images/trusted-signing-azure.png)
```

And this can also be done before a NuGet package has been signed. This copied value should then be entered into a form that adds
this EKU to the user. As mentioned earlier the Subject and Issuer will be set once the first package that has this EKU is uploaded.

#### How does this help the user?

Adding this in the interface would make it possible for a user to register an EKU in their account. This would be easier than
extracting the certificate from a signed NuGet package. The process to get that `.cer` file from a signed package is not that
simple. This would be the final step that would give a user full control over their EKU's.

### Technical explanation

<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->

## Drawbacks

<!-- Why should we not do this? -->

## Rationale and alternatives

> <!-- Why is this the best design compared to other designs? -->

Reviewer 3: Isn't an alternative solution that we integrate with Trusted Signing and retrieve the subscriber ID directly?

Author: That would still require some kind of integration with the Azure API that would require authorization setup. That feels more complicated?

Reviewer 3: I'm not suggesting this as a replacement for the current proposal.

The purpose of this section is to present alternatives and rationale "[w]hy this is the best design compared to other designs." Complexity and cost are solid motivators.

When someone looks back and wonders if we considered option ___, ideally, it would have been captured in this section.

Reviewer: I can also help with testing as I am also onboarded with Trusted Signing.

Reviewer 2: I think we should start with the proposal as-is. This sounds like a convenience to avoid the copy-paste of Trusted Signing details (or discovering it from a certificate).

The current proposal is sufficient for first step (or, first set of 3 steps 😀)

<!-- What other designs have been considered and why weren't they chosen? -->
<!-- What is the impact of not doing this? -->

## Prior Art

<!-- What prior art, both good and bad are related to this proposal? -->
<!-- Do other features exist in other ecosystems and what experience have their community had? -->
<!-- What lessons from other communities can we learn from? -->
<!-- Are there any resources that are relevant to this proposal? -->

## Unresolved Questions

> <!-- What parts of the proposal do you expect to resolve before this gets accepted? -->

Reviewer 3: How do we (NuGet.org) get a Trusted Signing certificate for testing this feature?

Reviewer 2: I can help with manual testing since I am personally onboarded. I am not sure if we need regular E2E testing (needing a new Trusted Signing signature per test run). Are you referring to manual or automated testing?

Reviewer 3: > I can help with manual testing since I am personally onboarded.

Suppose you go on leave for a while.

> Are you referring to manual or automated testing?

Neither specifically at this time, though automated testing is preferrable.  I'm thinking about E2E testability during development, release, and post-release.

Reviewer 2: This is work we can take on as the engineering team owning the effort. I agree with your point, however. We should invest in automation to make sure this keeps working over time.

automatically linking the Durable Identity Value to the user by using the current "certificate upload" interface. The user would need to do
the following:
- Extract the `.cer` file of the signing certificate from the NuGet package.
- Upload the `.cer` file and link it to their account.
- Upload the NuGet package that was signed with Trusted Signing.

Below is a picture that show how the certificate and Durable Identity Value would be linked to user before and after the NuGet package is
uploaded.

![Durable Identity Value link](images/trusted-signing-durable-identity-value-link.png)

This would mean that the certificate is no longer linked to the user and will no longer be shown in the interface because the
Durable Identity Value is now linked to the user. That will mean that the user would need to contact the NuGet team to remove
the Durable Identity Value link until updates to the interface have been made. But I don't think a lot of people will need this
functionality right away.

The Durable Identity Value will only be linked when the following conditions are true:
- The package was signed within the last 30 days.

> - The signing certificate must be issued by the root CA certificate `Microsoft Identity Verification Root Certificate Authority 2020`.

Reviewer 3: ```suggestion
- The signing certificate's issuer must chain to the `Microsoft ID Verified Code Signing PCA 2021` certificate that is issued by the root CA certificate `Microsoft Identity Verification Root Certificate Authority 2020`.
```

Author: I have updated this with your suggestion and also rebased the PR.
